### PR TITLE
Custom Reports optimization

### PR DESCRIPTION
**Pull request by carkom**
Custom Reports can be very slow for larger budget files. I've remade the data pulls to speed things up.

### CODE REVIEW

**Comment by carkom** [comment]
@MatissJanis, can you check that this PR loads your payees faster and functions more to your expectations?

**Comment by Kidglove57** [comment]
> @youngcw @Kidglove57
> 
> Can you check that this PR works as you'd expect with your budget files? Cheers!

Thank you so much @carkom. I just realised how much I needed this. I transfer loans, from a "for budget" account, to my adult kids and keep the resultant loan account off-budget. So this fix made me realise that my reporting had been all wrong before. Now it's fixed and I cannot see any issues.  Brilliant!

**Comment by MatissJanis** [comment]
Loading my payees is wayyyyyyy faster now! Good job.

**Comment by MatissJanis** [comment]
The next release is in just a few days so I'm quite worried about merging this PR as it has a high risk factor to it. Any objections if we hold off until the release is done?

**Comment by carkom** [comment]
> The next release is in just a few days so I'm quite worried about merging this PR as it has a high risk factor to it. Any objections if we hold off until the release is done?

Nope no worries to hold on this one.

**Comment by Kidglove57** [comment]
I would certainly have argued for this being merged now, for the following reasons:
1) It is experimental feature already and comes with quite a few warnings  - so the risk of merging is reduced 
2) It is a bug fix and not a new feature - without it the reports are wrong. Or at least where categorised transfers in are concerned.
3) December may be a critical month for users’ budgets as they review their category funding in preparation  for 2024. I particularly use the 12 month average to help determine my monthly category fill in the new year. So that figure needs to be accurate.

just an alternative few but ultimately I would need to defer to developer colleagues.

**Comment by MatissJanis** [comment]
@Paul: this is a relatively large Pr that touches files outside of custom
reports too. It is not limited to only custom reports. Which is why it’s
high-risk to merge it so close to the release.

Does that change your thoughts?

Matiss Janis Aboltins

**Comment by Kidglove57** [comment]
On the assumption that the risks of what this may touch outside of Custom Reports outweigh the risks of folk generating incorrect reports for themselves then yes, I do of course defer to you guys, until you feel it is ready.

**Comment by carkom** [comment]
The PR looks bigger than it actually is because about half the lines are renaming variables for clarity. Bulk of the actual code changes  are in 3 or 4 files in the spreadsheets folder. The only file outside the reports directory is a compiler addition for ids (4 lines of code).

**Comment by MatissJanis** [comment]
Ok, so I think in order to minimize the risk we have three options:

1. undo all the code reorganization/variable rename changes and focus only on the real bugfix in this PR and scope it to only the custom reports files - thus decreasing the blast radius in other parts of the product and reducing the risk
2. add an extra warning message to the custom reports page that "this is highly experimental feature; the number might be inaccurate" etc. thus minimizing the risk of people generating incorrect reports and getting confused
3. do nothing for the current release; the custom reports feature is already under a feature flag that warns about unstable features, so folks need to expect that it will not be perfect on day 1; we can merge this straight after the release so folks can jump to edge if they want to use a more stable version of custom reports

If (1) is possible - that sounds like the best way to go. But I don't know how technically feasible it is to reduce the scope of this PR. My gut feel is that it won't be simple.

Or maybe there is a third option? As long as we don't merge a big and risky PR just under a day before the release - I'll be happy. I _really_ don't want a repeat of publishing patches a day after the release goes out. That just makes the product look unstable.

**Comment by carkom** [comment]
I'd agree with that. Option (3) seems the best, stick to our processes - it is experimental after all, majority of users won't have it enabled anyway.

File: packages/desktop-client/src/components/reports/ReportTableHeader.tsx
Comment (carkom) [inline packages/desktop-client/src/components/reports/ReportTableHeader.tsx:43]: No changes to this code just tabbed over for formatting

File: packages/desktop-client/src/components/reports/ReportTableTotals.tsx
Comment (carkom) [inline packages/desktop-client/src/components/reports/ReportTableTotals.tsx:41]: No changes to code just tabbed for formatting

File: packages/desktop-client/src/components/reports/spreadsheets/makeQuery.tsx
Comment (carkom) [inline packages/desktop-client/src/components/reports/spreadsheets/makeQuery.tsx:2]: No big changes here, just split to it's own file.

File: packages/desktop-client/src/components/reports/spreadsheets/recalculate.tsx
Comment (carkom) [inline packages/desktop-client/src/components/reports/spreadsheets/recalculate.tsx:2]: Same as above, no big changes just saved to a separate file.

**Comment by carkom** [comment]
@MatissJanis this is ready for review. I have another PR which I'll submit now which goes and fixes all the typeScript files with type definitions and adds new (often used) entities. I've commented in line (above) any call outs. Cheers!

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:48]: **💬 suggestion:** ‏do we have a type we could use to enforce all fields being correctly set here? If not - can we create it? Same for the others

```suggestion
let uncategorizedCategory: Something = {
```

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:83]: **❓ question:** ‏what's `uncat`? Variable naming could be improved.

Also missing function argument types.

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:104]: **💬 suggestion:** ‏should add argument types
```suggestion
export const groupBySelections = (
  groupBy: Something,
  categoryList,
  categoryGroup,
  payees,
  accounts,
) => {
```

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:132]: **💬 suggestion:** ‏would be good to add a fatal error here to stop program execution if this condition is ever met (but I hope it is never met).
```suggestion
    default:
        throw new Error('Something');
```

File: packages/desktop-client/src/components/reports/graphs/BarGraph.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/graphs/BarGraph.tsx:133]: **👏 praise:** ‏thanks for adding missing types

File: packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.tsx:1]: **💬 suggestion:** ‏two things here:

1. missing argument types
2. the filename should be `.ts` instead of `.tsx` (`.tsx` implies using JSX; i.e. - a component. Whereas this is just a util function)

File: packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.tsx:7]: **💭 thought:** ‏this expression is very difficult to read. Is there anything we could do to improve the readability of it? Perhaps breaking it down into smaller conditional statements that return early?

i.e.
```
if (something) {
  return true;
}

if (somethingElse) {
   return false;
}

return true;
```

File: upcoming-release-notes/1988.md
Comment (MatissJanis) [inline upcoming-release-notes/1988.md:6]: **💭 thought:** ‏people reading the changelog won't understand what you've actually done in this change. It would be nice if you could write out what this PR is improving. For example: `Data loading performance improvements for custom reports`. The current "optimizations" is very vague.

File: packages/desktop-client/src/components/reports/spreadsheets/recalculate.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/recalculate.tsx:22]: **🥜 nitpick:** ‏
```suggestion
```

File: packages/desktop-client/src/components/reports/spreadsheets/recalculate.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/recalculate.tsx:2]: **💬 suggestion:** ‏this file should also be `.ts`

File: packages/desktop-client/src/components/reports/spreadsheets/makeQuery.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/makeQuery.tsx:2]: **💬 suggestion:** ‏this should also be `.ts`

File: packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.tsx:11]: **💬 suggestion:** ‏this should also be `.ts`

File: packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.tsx:12]: **💬 suggestion:** ‏missing argument types

**Review by MatissJanis — CHANGES_REQUESTED** [review]
Thanks for your hard work on this!

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (carkom) [reply packages/desktop-client/src/components/reports/ReportOptions.tsx:83]: varibale naming was all moved to #1996 

Argument types added to entire file.

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:73]: **💬 suggestion:** ‏
```suggestion
export type UncategorizedEntity = CategoryEntity & {
  id: null;
  uncat_id: string;
  offBudget: boolean; // what does this do? naming could be improved; also: we have a mixture of `camelCase` and `snake_case` here - ideally should use just one
  transfer: boolean; // what does this do? naming could be improved
  category: boolean; // what does this do? naming could be improved
};
```

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:80]: **💬 suggestion:** ‏
```suggestion
export type UncategorizedGroupEntity = CategoryGroupEntity & {
  id: null;
  categories?: UncategorizedEntity[];
};
```

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:124]: **❓ question:** ‏what's `uncat`? Variable naming could be improved

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/ReportOptions.tsx:121]: **💬 suggestion:** ‏these are not really `props`. Component have `props`. Utility functions have `arguments/parameters`. 

Also: this is a total nitpick, but I'd recommend just inlining the type definition for utility functions. Makes your life easier as you won't need to create a named type.

File: packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.ts
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.ts:20]: **💬 suggestion:** ‏
```suggestion
  return data.filter(
      asset => {
          if (!item.uncat_id) {
              return true;
          }
          
          // TODO: variable naming could definitely be improved; I just came up with some example vars
          const isTransferableAccount = (item.transfer ? asset.transferAccount : !asset.transferAccount);
          const shouldItemBeHidden = (item.category ? true : !asset.category);
          const isOffBudget = (item.offBudget ? asset.accountOffBudget : !asset.accountOffBudget);
          
          return isTransferableAccount && shouldItemBeHidden && isOffBudget;
      }
  );
```

File: packages/desktop-client/src/components/reports/spreadsheets/default-spreadsheet.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/default-spreadsheet.tsx:61]: **🔨 warning:** ‏TS should be able to automatically infer the return type of the function. If it's not doing so - this will likely point to an issue in the implementation.

```suggestion
  let [catList, catGroup] = categoryLists({ hidden, uncat, categories });
```

(changing the return statement of `categoryList` to `return [something, something] as const` might solve the problem for you)

File: packages/desktop-client/src/components/reports/spreadsheets/default-spreadsheet.tsx
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/default-spreadsheet.tsx:93]: **🔨 warning:** ‏same problem here
```suggestion
    const [assets, debts] = await Promise.all([
```

File: packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.ts
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.ts:34]: **🔨 warning:** ‏same problem here

File: packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.ts
Comment (MatissJanis) [inline packages/desktop-client/src/components/reports/spreadsheets/grouped-spreadsheet.ts:55]: This is good 👏

**Review by MatissJanis — CHANGES_REQUESTED** [review]

File: packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.tsx
Comment (MatissJanis) [reply packages/desktop-client/src/components/reports/spreadsheets/filterHiddenItems.tsx]: Made a code suggestion here: https://github.com/actualbudget/actual/pull/1988#discussion_r1414326875

File: packages/desktop-client/src/components/reports/ReportOptions.tsx
Comment (MatissJanis) [reply packages/desktop-client/src/components/reports/ReportOptions.tsx:124]: Would you be open to addressing the feedback in this PR instead? I really don't want to be jumping back-and-forth between multiple PRs for a single review.

Comment (carkom) [reply packages/desktop-client/src/components/reports/ReportOptions.tsx:124]: Renaming all these categories will nearly double the size of this PR. As previous discussed, this PR is already fairly large. I don't see the need to redo work I've already done. These variable names are not being introduced in this PR, they were already here from the original PR.

You don't need to jump back and forth between the PRs. Just don't worry about variable names for this PR as they will all be changed in an upcoming PR. My apologies for not making them better in the first place but I think the PR will become unmangeable if we try to add everything to it all at once.

Comment (MatissJanis) [reply packages/desktop-client/src/components/reports/ReportOptions.tsx]: Pardon my ignorance, but I don't see `categoryLists` on `master` branch, so it is indeed new code. It's not existing functionality that's just moving to a different location.

As such: it gets the full scrutiny of a proper code review.

And just to clarify: I don't expect you to change `uncat` variable in the entire codebase. Just this new util function that you are introducing.

Comment (carkom) [reply packages/desktop-client/src/components/reports/ReportOptions.tsx]: No no, that's on me. I'd assumed you meant changing them for the entire code base, sorry for that. I've updated them and used the new variable names to match the 1996 PR for these variables.

**Review by MatissJanis — APPROVED** [review]
Perfect! 🚀

**Review by MatissJanis — APPROVED** [review]

**Pull request merged**
